api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,656 @@
+	for i, resource := range resources {
+		new := models.Resource{
+			ID:     resource.ID,
+			InUse:  resource.InUse,
+			Ports:  resource.Ports,
+			Memory: resource.Memory,
+			CPU:    resource.CPU,
+		}
+		resourceModels[i] = new
+	}
+
+	return resourceModels
+}
+
+func (e *EnvironmentScaler) getResourceProviders(clusterName string) ([]*models.Resource, error) {
+	listContainerInstancesInput := &ecs.ListContainerInstancesInput{}
+	listContainerInstancesInput.SetCluster(clusterName)
+	listContainerInstancesInput.SetStatus("ACTIVE")
+
+	containerInstanceARNs := []*string{}
+	listContainerInstancesPagesFN := func(output *ecs.ListContainerInstancesOutput, lastPage bool) bool {
+		containerInstanceARNs = append(containerInstanceARNs, output.ContainerInstanceArns...)
+
+		return !lastPage
+	}
+
+	if err := e.Client.ECS.ListContainerInstancesPages(listContainerInstancesInput, listContainerInstancesPagesFN); err != nil {
+		return nil, err
+	}
+
+	describeContainerInstancesInput := &ecs.DescribeContainerInstancesInput{}
+	describeContainerInstancesInput.SetCluster(clusterName)
+	describeContainerInstancesInput.SetContainerInstances(containerInstanceARNs)
+
+	output, err := e.Client.ECS.DescribeContainerInstances(describeContainerInstancesInput)
+	if err != nil {
+		return nil, err
+	}
+
+	result := []*models.Resource{}
+	if len(containerInstanceARNs) == 0 {
+		return result, nil
+	}
+
+	if len(output.ContainerInstances) == 0 {
+		return result, nil
+	}
+
+	for _, instance := range output.ContainerInstances {
+		// it's non-intuitive, but the ports being used by the tasks live in
+		// instance.RemainingResources, not instance.RegisteredResources
+		var (
+			usedPorts       []int
+			availableMemory bytesize.Bytesize
+			availableCPU    bytesize.Bytesize
+		)
+
+		for _, resource := range instance.RemainingResources {
+			fmt.Println(resource)
+			switch aws.StringValue(resource.Name) {
+			case "MEMORY":
+				val := aws.Int64Value(resource.IntegerValue)
+				availableMemory = bytesize.MiB * bytesize.Bytesize(val)
+
+			case "PORTS":
+				for _, port := range resource.StringSetValue {
+					port, err := strconv.Atoi(aws.StringValue(port))
+					if err != nil {
+						log.Printf("[WARN] Instance %s: Failed to convert port to int: %v", aws.StringValue(instance.Ec2InstanceId), err)
+						continue
+					}
+
+					usedPorts = append(usedPorts, port)
+				}
+			case "CPU":
+				val := aws.Int64Value(resource.IntegerValue)
+				availableCPU = bytesize.MiB * bytesize.Bytesize(val)
+			}
+		}
+
+		inUse := aws.Int64Value(instance.PendingTasksCount)+aws.Int64Value(instance.RunningTasksCount) > 0
+		r := &models.Resource{
+			ID:     aws.StringValue(instance.Ec2InstanceId),
+			InUse:  inUse,
+			Ports:  usedPorts,
+			Memory: availableMemory.Megabytes(),
+			CPU:    availableCPU.Megabytes(),
+		}
+
+		result = append(result, r)
+	}
+
+	return result, nil
+}
+
+func (e *EnvironmentScaler) getResourceConsumers(clusterName string) ([]*models.Resource, error) {
```

Comment:
This function needs to be broken up

Reply:
This will be broken up into three functions as a result of #440.